### PR TITLE
clang-tidy: Apply modernize-loop-convert everywhere

### PR DESCRIPTION
**Pull request by sakertooth**
Extremely useful here in my opinion.

This clang-tidy check transforms for loops into their range-based loop equivalent where applicable. This in particular was ran using the ``MinConfidence`` option set to reasonable, which is the default.

### CODE REVIEW

File: plugins/GigPlayer/GigPlayer.cpp [hunk 1]
```diff
@@ -408,17 +407,17 @@ void GigInstrument::play( sampleFrame * _working_buffer )
...
+		if( !( m_note.state == PlayingKeyDown ||
+				m_note.state == PlayingKeyUp ) )
 		{
 			continue;
 		}
 
-		for( QList<GigSample>::iterator sample = it->samples.begin();
-				sample != it->samples.end(); ++sample )
+		for( QList<GigSample>::iterator sample = m_note.samples.begin();
+				sample != m_note.samples.end(); ++sample )
```
Comment (DomClark) [inline plugins/GigPlayer/GigPlayer.cpp:420]: Interestingly, it hasn't converted this loop. Perhaps this is because it was already changed by converting the outer loop, in which case multiple passes may be required.

File: plugins/LadspaEffect/LadspaControlDialog.cpp [hunk 2]
```diff
@@ -70,10 +70,9 @@ LadspaControlDialog::LadspaControlDialog( LadspaControls * _ctl ) :
 void LadspaControlDialog::updateEffectView( LadspaControls * _ctl )
 {
 	QList<QGroupBox *> list = findChildren<QGroupBox *>();
-	for( QList<QGroupBox *>::iterator it = list.begin(); it != list.end();
-									++it )
+	for(auto & it : list)
```
Comment (DomClark) [inline plugins/LadspaEffect/LadspaControlDialog.cpp:73]: When the container isn't a plural word, it seems to reuse the iterator name from before. This is misleading, as the variable is no longer an iterator, but the item itself.

File: src/core/ConfigManager.cpp [hunk 3]
```diff
@@ -609,25 +607,22 @@ void ConfigManager::saveConfigFile()
 	lmms_config.setAttribute("configversion", m_configVersion);
 	doc.appendChild(lmms_config);
 
-	for(settingsMap::iterator it = m_settings.begin();
```
[the remark this answers is not in the document]
Comment (DomClark) [reply src/core/ConfigManager.cpp:612]: Yeah, I wasn't suggesting you change it here. I was wondering why it hadn't been changed automatically, then I realised you can't access a `QMap`'s key using a range-based for loop.

File: plugins/Lv2Instrument/Lv2Instrument.cpp [hunk 4]
```diff
@@ -79,7 +79,7 @@ Lv2Instrument::Lv2Instrument(InstrumentTrack *instrumentTrackArg,
 	if (Lv2ControlBase::isValid())
 	{
 #ifdef LV2_INSTRUMENT_USE_MIDI
-		for (int i = 0; i < NumKeys; ++i) { m_runningNotes[i] = 0; }
+		for (int& m_runningNote : m_runningNotes) { m_runningNote = 0; }
```
[the remark this answers is not in the document]
Comment (sakertooth) [reply plugins/Lv2Instrument/Lv2Instrument.cpp:82]: Even simpler would be making ``m_runningNotes`` a ``std::array`` and using ``m_runningNotes.fill(0);``, but yeah like you said it is probably out of scope for this PR.

File: src/core/AutomationClip.cpp [hunk 5]
```diff
@@ -1004,17 +1000,14 @@ void AutomationClip::resolveAllIDs()
...
-			for( Track::clipVector::iterator j = v.begin();
-							j != v.end(); ++j )
+			Track::clipVector v = track->getClips();
+			for (auto& clip : v)
 			{
-				AutomationClip * a = dynamic_cast<AutomationClip *>( *j );
+				AutomationClip* a = dynamic_cast<AutomationClip*>(clip);
 				if( a )
 				{
 					for( QVector<jo_id_t>::Iterator k = a->m_idsToResolve.begin();
```
[the remark this answers is not in the document]
Comment (sakertooth) [reply src/core/AutomationClip.cpp:1013]: What happened most likely was that I forgot to add ``--format-style=file`` the first time around, so when it ran it didn't consider formatting the changes :sweat_smile:. I could've just reran it with the format style set, but I figured I would do it manually because there were more issues you pointed out that needed to be addressed than just the formatting that clang-tidy wouldn't resolve (i.e, m_ prefix removal, clang-tidy not making the optimal variable name change within the range loop, etc).

File: src/core/Song.cpp [hunk 6]
```diff
@@ -99,7 +99,7 @@ Song::Song() :
 	m_loopRenderRemaining(1),
 	m_oldAutomatedValues()
 {
-	for(int i = 0; i < Mode_Count; ++i) m_elapsedMilliSeconds[i] = 0;
+	for (double& elapsedMilliSecond : m_elapsedMilliSeconds) elapsedMilliSecond = 0;
```
[the remark this answers is not in the document]
Comment (sakertooth) [reply src/core/Song.cpp:102]: Switching it to ``millisecondsElapsed``, which should be satisfactory?

File: src/core/AutomationClip.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (DomClark) [reply src/core/AutomationClip.cpp]: Sounds reasonable to me. I only spotted this one because the top of the loop showed up in the diff; I bet there are a bunch more out there that it hasn't touched. It's probably not worth scouring the code for every last one, so I won't worry about it further.

**Review by DomClark — APPROVED** [review]

File: plugins/MidiExport/MidiExport.cpp [hunk 7]
```diff
@@ -216,27 +216,27 @@ bool MidiExport::tryExport(const TrackContainer::TrackList &tracks,
 				int len = n.toElement().attribute("steps", "1").toInt() * 12;
 
 				// for each pattern clip of the current pattern track (in song editor)
-				for (auto it = plist.begin(); it != plist.end(); ++it)
+				for (auto& pair : plist)
```
[the remark this answers is not in the document]
Comment (allejok96) [reply plugins/MidiExport/MidiExport.cpp:219]: Or maybe it's time for another

    for (auto&& [start, end] : plist)

Which I don't understand half of even after reading [Dom's explanation](https://discord.com/channels/203559236729438208/784594576223633478/1012474458956640339) but it looks cool and dangerous

File: src/core/Mixer.cpp [hunk 8]
```diff
@@ -388,14 +388,13 @@ void Mixer::moveChannelLeft( int index )
 	TrackContainer::TrackList patternTrackList = Engine::patternStore()->tracks();
 
 	TrackContainer::TrackList trackLists[] = {songTrackList, patternTrackList};
-	for(int tl=0; tl<2; ++tl)
+	for (auto& trackList : trackLists)
```
Comment (allejok96) [inline src/core/Mixer.cpp:391]: perhaps?
```suggestion
	for (auto& trackList : {songTrackList, patternTrackList})
```

File: src/gui/MixerView.cpp [hunk 9]
```diff
@@ -240,20 +240,19 @@ void MixerView::updateMaxChannelSelector()
 	TrackContainer::TrackList patternStoreTracks = Engine::patternStore()->tracks();
 
 	TrackContainer::TrackList trackLists[] = {songTracks, patternStoreTracks};
-	for(int tl=0; tl<2; ++tl)
+	for (auto& trackList : trackLists)
```
Comment (allejok96) [inline src/gui/MixerView.cpp:243]:
```suggestion
	for (auto& trackList : {songTracks, patternStoreTracks})
```

File: src/tracks/PatternTrack.cpp [hunk 10]
```diff
@@ -185,11 +184,9 @@ void PatternTrack::loadTrackSpecificSettings(const QDomElement& _this)
 					Engine::patternStore()->tracks();
 		// copy clips of all tracks from source pattern (at bar "src") to destination
 		// clips (which are created if they do not exist yet)
-		for( TrackContainer::TrackList::iterator it = tl.begin();
-							it != tl.end(); ++it )
+		for (auto& track : tl)
```
Comment (allejok96) [inline src/tracks/PatternTrack.cpp:187]:
```suggestion
		for (auto& track : Engine::patternStore()->tracks())
```

File: src/core/Piano.cpp [hunk 11]
```diff
@@ -67,9 +67,9 @@ Piano::Piano( InstrumentTrack* track ) :
 	m_instrumentTrack( track ),
 	m_midiEvProc( track )        /*!< the InstrumentTrack Model */
 {
-	for( int i = 0; i < NumKeys; ++i )
+	for (bool& pressedKey : m_pressedKeys)
```
Comment (allejok96) [inline src/core/Piano.cpp:70]: Another case for `std::array` and `fill`

**Review by allejok96 — APPROVED** [review]
Everything looks good, just have some optional suggestions.

File: plugins/MidiExport/MidiExport.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (sakertooth) [reply plugins/MidiExport/MidiExport.cpp:219]: > Or maybe it's time for another
> 
> ```
> for (auto&& [start, end] : plist)
> ```
> 
> Which I don't understand half of even after reading [Dom's explanation](https://discord.com/channels/203559236729438208/784594576223633478/1012474458956640339) but it looks cool and dangerous

Yeah, I'm somewhat unsure as well, but I really think it depends on context and what the ``range-expression`` is (in this case ``plist``). I've been doing some reading, and from what I understand, ``auto&&`` (r-value references) only bind to r-values, ``auto&`` (l-value reference) only bind to l-values, and ``const auto&`` (const l-value reference) can bind to both l-values and r-values.

So, I'm guessing that the most correct choice of using ``auto&``, ``auto&&``, or ``const auto&`` depends on what kind of value ``range-expression`` returns. It's important to note that other value categories, such as glvalues and xvalues, fall under either an l-value or an r-value, so they apply as well.

In this case, if ``range-expression`` (``plist``) gives back an l-value, ``auto&`` or ``const auto&`` can be used. If ``range-expression`` gives back an r-value, ``auto&&`` or ``const auto&`` can be used. So if the expression is not changed given the range-based for loop (i.e it can be const), ``const auto &`` may be the most optimal choice given a wide variety of circumstances. 

However, it is possible to convert an l-value to an r-value given an implicit l-value to r-value conversion. Therefore, this would allow ``auto&&`` to bind to both l-values and r-lvaues.

So ultimately, I think the main benefit of using ``auto&&`` in range-based for loops is if you need to modify the value in question, but also need it to bind to both l-values and r-values.

Structured bindings don't really change this idea at all. The type of a structured binding is compiler generated (hence why we use ``auto`` for them), but at the end of the day that just means it's a ``range-declaration`` that has one is like any other declaration.

Given my current understanding of this, I think I will update my changes to use ``const auto&`` where applicable, and ``auto&&`` otherwise.

**TLDR: If you don't need to modify the range declaration, ``const auto&`` should be used. If you do need to modify the range declaration, ``auto&&`` should be used.**

File: src/core/Piano.cpp (hunk 11 above)
Comment (sakertooth) [reply src/core/Piano.cpp:70]: Could probably simplify this out completely with value initialization though.

File: plugins/MidiExport/MidiExport.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (PhysSong) [reply plugins/MidiExport/MidiExport.cpp]: > from what I understand, `auto&&` (r-value references) only bind to r-values

It can bind to both `T&`(`auto` becomes `T&`) and `T&&`(`auto` becomes `T`).

> If you don't need to modify the range declaration, `const auto&` should be used. If you do need to modify the range declaration, `auto&&` should be used.

That sounds reasonable, indeed. I would use `T&` in case I need to both modify the declaration and make sure it's bound to an lvalue, BTW.

[the remark this answers is not in the document]
Comment (sakertooth) [reply plugins/MidiExport/MidiExport.cpp]: > That sounds reasonable, indeed. I would use `T&` in case I need to both modify the declaration and make sure it's bound to an lvalue, BTW.

Yeah, I'll add that if the range-based for loop can go without needing to bind to r-values, a ``auto&`` is a nice choice mainly due to simplicity and readability reasons. 

...However, when seeing a range-based for loop, the compiler would generate something like this (from https://en.cppreference.com/w/cpp/language/range-for):

```cpp
    auto && __range = range-expression ;
    auto __begin = begin-expr ;
    auto __end = end-expr ;
    for ( ; __begin != __end; ++__begin)
    {

        range-declaration = *__begin;
        loop-statement

    }
```
And given ``range-declaration = *__begin;``, I read that the indirection operator ``*`` gives back an l-value, or more specifically a ``T&``. Then again, it can be overloaded, and in cases like ``std::vector<bool>``, dereferencing the result of ``std::vector<bool>::begin`` doesn't return an l-value, but a temporary r-value with type ``std::vector<bool>::reference``, as seen here (in particular with the GCC compiler).

```cpp
_GLIBCXX_NODISCARD _GLIBCXX20_CONSTEXPR
reference
operator*() const
{ return reference(_M_p, 1UL << _M_offset); }
 ```

Which cannot bind to an l-value reference, such as ``auto&``.

Okay, I think I got it now. As you stated, ``auto&`` should be used if it is sure to be bound to an l-value, which the indirection operator **usually** returns, but not always. This would then pave way for the use of ``auto&&``, but if you don't need to modify anything, ``const auto&`` works way better in terms of intent, simplicity, and readability.

[the remark this answers is not in the document]
Comment (PhysSong) [reply plugins/MidiExport/MidiExport.cpp]: Good point for rare edge cases such as `std::vector<bool>`, which I never used.

**Comment by sakertooth** [comment]
Okay, I think this is good now. Made a few substantial adjustments, but this PR should be clear for merging. One final review would help, if anyone is up for it, but it's probably not necessary.

**Comment by JohannesLorenz** [comment]
I originally considered reviewing this, but it already has 2 reviews, which should be enough.

@sakertooth Can you please fix the merge conflicts?

**Comment by sakertooth** [comment]
> I originally considered reviewing this, but it already has 2 reviews, which should be enough.
> 
> @sakertooth Can you please fix the merge conflicts?

Will do.
